File: MyLeetCode/python/Range Addition.py
###
# Array
# Time Complexity: hasNext:  O(k+n) k:operations, n: length
# Space Complexity: O(1)
# Note. if the elements are not the same, change the offset of the element:
# [2,3,10,5] transforms to [2, 1, 7, -5]
###
class Solution(object):
    def getModifiedArray(self, length, updates):
        """
        :type length: int
        :type updates: List[List[int]]
        :rtype: List[int]
        """
        re = [0] * length
        for s, e, v in updates:
            re[s] += v
            if e + 1 < length:
                re[e+1] -= v
        for i in xrange(1,length):
            re[i] += re[i-1]
        return re
                                        
                            
                    

File: MyLeetCode/Leetcode/Range Addition.cpp
//update the start index by add the val and make an offset to the end+1 index 
//Time complexity:O(n+k)
class Solution {
public:
    vector<int> getModifiedArray(int length, vector<vector<int>>& updates) {
        vector<int> res(length, 0);
        for (auto e : updates) {
            int start = e[0];
            int end = e[1];
            int val = e[2];
            res[start] += val;
            if (end + 1 < length) {
                res[end + 1] -= val;
            }
        }
        
        for (int i = 1; i < length; i++) {
            res[i] += res[i-1];
        }
        return res;
    }
};